doc/port_link_training/port-link-training-design.md

```diff
@@ -0,0 +1,369 @@
+- Port configuration setting flow will be changed in orchagent of sonic-swss. See detail description in section [SWSS Enhancements](#swss-enhancements).
+
+### SAI API Requirement
+
+First of all, we'll leverage the existing SAI port attributes listed below  
+
+```cpp
+    /**
+     * @brief Enable/Disable Port Link Training
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_PORT_ATTR_LINK_TRAINING_ENABLE,
+
+    /**
+     * @brief Link training failure status and error codes
+     *
+     * @type sai_port_link_training_failure_status_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_LINK_TRAINING_FAILURE_STATUS,
+
+    /**
+     * @brief Status whether the receiver trained or not trained to receive data
+     *
+     * @type sai_port_link_training_rx_status_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_LINK_TRAINING_RX_STATUS,
+```
+
+On the other hand, a new port attribute will be introduced for the link-training ability query, and please note the link-training ability could be unavailable on the certain ports on a switch silicon (e.g. The uplink/management ports), hence we need to do the query at per-port basis.  
+
+```cpp
+    /**
+     * @brief Query link-training support
+     *
+     * @type bool
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_SUPPORTED_LINK_TRAINING_MODE,
+```
+
+Vendor-specific SAI implementation is not in the scope of this document, but there are some common requirements for SAI:
+
+1. swss must check the link-training abilities before making the corresponding request to syncd.
+2. SAI implementation must return error code if any of the above attributes is not supported, swss and syncd must not crash.
+3. SAI implementation must keep port link training disabled by default for backward compatibility. As long as swss and SAI keep backward compatible, user need not change anything after this feature is implemented and available in SONiC.
+
+### Configuration and management 
+
+#### CLI Enhancements
+
+A few new CLI commands are designed to support port link training.
+
+##### Config link training mode
+
+
+```
+Format:
+  config interface link-training <interface_name> <mode> [-f]
+
+Arguments:
+  interface_name: [mandatory] name of the interface to be configured. e.g: Ethernet0
+  mode: [mandatory] link training mode, can be either "on" or "off"
+  -f: [optional] forcing link-training configuration on an unsupported transceiver
+
+Example:
+  config interface link-training Ethernet0 on
+  config interface link-training Ethernet0 on -f
+  config interface link-training Ethernet0 off
+
+Return:
+  - error message if interface_name or mode is invalid
+  - error message if the link-training is not supported on the interface
+  - empty upon success
+```
+
+For deployment considerations, when the link-training is enabled on a transceiver without LT capabilities, this config command should report an error and get aborted, unless '-f' option is specified.
```

Comment:
Since we have removed pmon related sections, I believe this line becomes irrelevant. Can you please remove?

Reply:
HLD updated